libfido2: 1.10.0 -> 1.11.0 and refactor to enable NFC support via pcsc

bootstrap-prime opened the pull request:
###### Description of changes

libfido2 supports nfc via pcsc, which could be used to let ssh access a yubikey through an NFC reader. This PR refactors dbus, polkit, python-dbusmock, bluez and pcsclite to remove a dependency loop on systemd, and updates libfido2 to the latest version.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

bootstrap-prime committed:
libfido2: 1.10.0 -> 1.11.0

bootstrap-prime commented:
I actually can't reproduce this ofborg failure, nix-build -A aws-sdk-cpp builds successfully for me.

bootstrap-prime commented:
also requesting review by @prusnak and @dtzWill, maintainers of libfido2

marsam commented:
@GrahamcOfBorg eval

prusnak reviewed (approved):
ACK libfido2 changes

The rest of changes look OK too, but I am no expert in systemd packaging

prusnak commented:
@jtojnar can you please review freedesktop stuff? (changes to dbus+polkit)

jtojnar commented on pkgs/development/libraries/libfido2/default.nix:
```diff
@@ -8,23 +8,24 @@
   buildInputs = [ libcbor openssl zlib ]
     ++ lib.optionals stdenv.isDarwin [ hidapi ]
-    ++ lib.optionals stdenv.isLinux [ udev ];
+    ++ lib.optionals stdenv.isLinux [ udev pcsclite.dev ];
```
The `dev` output should be the default: https://nixos.org/manual/nixpkgs/stable/#sec-multiple-outputs-using-split-packages

jtojnar commented on pkgs/development/python-modules/python-dbusmock/default.nix:
```diff
@@ -1,6 +1,7 @@
   nose, dbus, dbus-python, pygobject3,
-  which, pyflakes, pycodestyle, bluez, networkmanager
+  which, pyflakes, pycodestyle, bluez, networkmanager,
+  doCheck ? true
```
I am not really fond of adding more control flags as parameters. I would prefer relying on https://github.com/NixOS/nixpkgs/pull/119942, if it works with `buildPythonPackage`.

jtojnar reviewed (commented):
Would be nice to separate the `bluez`, `pcslite`, `python-dbusmock`, `dbus` and `polkit` changes into a separate commits, since they are independent from each other and from the `libfido2` ones.

jtojnar commented on pkgs/development/python-modules/python-dbusmock/default.nix:
```diff
@@ -53,11 +54,13 @@ buildPythonPackage rec {
   checkPhase = ''
     runHook preCheck
-    export PATH="$PATH:${lib.getOutput "test" bluez}/test";
+    ${if doCheck then ''export PATH="$PATH:${lib.getOutput "test" bluez}/test'' else ""}";
```
Something like the following (untested) will be even cleaner IMO:

```patch
diff --git a/pkgs/development/libraries/polkit/default.nix b/pkgs/development/libraries/polkit/default.nix
index e9e3d94327b..9e600ee9f19 100644
--- a/pkgs/development/libraries/polkit/default.nix
+++ b/pkgs/development/libraries/polkit/default.nix
@@ -99,8 +99,10 @@ stdenv.mkDerivation rec {
     rsync
     (python3.withPackages (pp: with pp; [
       dbus-python
-      # Avoid dependency cycle.
-      (python-dbusmock.override { doCheck = false; })
+      (python-dbusmock.overridePythonAttrs {
+        # Avoid dependency cycle.
+        doCheck = false;
+      })
     ]))
 
     # man pages
diff --git a/pkgs/development/python-modules/python-dbusmock/default.nix b/pkgs/development/python-modules/python-dbusmock/default.nix
index ae5db2aebfe..871c25c4909 100644
--- a/pkgs/development/python-modules/python-dbusmock/default.nix
+++ b/pkgs/development/python-modules/python-dbusmock/default.nix
@@ -1,16 +1,22 @@
 { lib, buildPythonPackage, fetchFromGitHub, runtimeShell,
   nose, dbus, dbus-python, pygobject3,
   which, pyflakes, pycodestyle, bluez, networkmanager,
-  doCheck ? true
+  runCommand
 }:
 
+let
+  pbap-client = runCommand "pbap-client" { } ''
+    mkdir -p "$out/bin"
+    ln -s "${bluez.test}/test/pbap-client" "$out/bin/pbap-client"
+  '';
+in
 buildPythonPackage rec {
   pname = "python-dbusmock";
   version = "0.26.1";
 
   src = fetchFromGitHub {
     owner = "martinpitt";
-    repo = pname;
+    repo = "python-dbusmock";
     rev = version;
     sha256 = "sha256-kavbWMTgKU/rBIo7RMs9NkwReYQyEdeFwMBSzEM9wa0=";
   };
@@ -49,18 +55,17 @@ buildPythonPackage rec {
 
   checkInputs = [
     nose dbus dbus-python which pycodestyle pyflakes
-    pygobject3 bluez (lib.getOutput "test" bluez) networkmanager
+    pygobject3 bluez
+    pbap-client
+    networkmanager
   ];
 
   checkPhase = ''
     runHook preCheck
-    ${if doCheck then ''export PATH="$PATH:${lib.getOutput "test" bluez}/test'' else ""}";
     nosetests -v
     runHook postCheck
   '';
 
-  inherit doCheck;
-
   meta = with lib; {
     description = "Mock D-Bus objects for tests";
     homepage = "https://github.com/martinpitt/python-dbusmock";
```

bootstrap-prime replied:
Looks good, this builds for me!

bootstrap-prime replied:
ofborg is unhappy with this, should I use the original test attribute acquisition method? 
`lib.getOutput "test" bluez`

jtojnar replied:
That would fix evaluation by just falling back to `out` output, which does not contain the file.

I guess the issue is that we are limiting the test output to `x86_64`:

https://github.com/NixOS/nixpkgs/blob/025f00750d07df498b42bd6ebea38c6802fe9857/pkgs/os-specific/linux/bluez/default.nix#L108

Removing the output was introduced in:

https://github.com/NixOS/nixpkgs/commit/0bbdba2d119caa0fccf48529cf374df7075403d2

but it does not make much sense to me.

I would probably make installing the `test` output unconditional.

bootstrap-prime committed:
polkit: use systemdMinimal to avoid dep cycle

bootstrap-prime committed:
pcsclite: use systemdMinimal to avoid dep cycle

bootstrap-prime committed:
bluez: use systemdMinimal to avoid dep cycle

also make test output unconditional

bootstrap-prime committed:
dbus: use systemdMinimal to avoid dep cycle

bootstrap-prime committed:
python-dbusmock: refactor check, avoid dep cycle

python-dbusmock relies on bluez for it's checks. Which shouldn't be a
problem, and isn't one normally. However, checkPhase contained a string
substitution that was always constructed regardless of the value of
doCheck (and simply not used if doCheck was false), and so bluez, which
was not supposed to be a dependency, was used as and registered as one
and caused a dependency cycle when trying to add pcsclite as a
dependency of libfido2.

The string substitution has been removed in favor of something that can
remain disabled when doCheck is false.

bootstrap-prime committed:
libfido2: depend on pcsclite to support nfc keys

jtojnar commented on pkgs/os-specific/linux/bluez/default.nix:
```diff
@@ -48,7 +48,7 @@ in stdenv.mkDerivation rec {
     python3.pkgs.wrapPython
   ];
 
-  outputs = [ "out" "dev" ] ++ lib.optional doCheck "test";
```
This will require corresponding changes to `postInstall`.

bootstrap-prime commented:
I can't reproduce this failure, so I'm going to request a re-evaluation.
@GrahamcOfBorg eval

bootstrap-prime commented:
Excellent, looks like the only thing left is failure to build on darwin. I'm not sure why bison is being built or timing out.

jtojnar reviewed (approved):
Diff looks good to me, did not test.

jiegec commented on pkgs/development/python-modules/python-dbusmock/default.nix:
```diff
@@ -1,15 +1,21 @@
   src = fetchFromGitHub {
     owner = "martinpitt";
-    repo = pname;
+    repo = "python-dbusmock";
```
Is this intended?

bootstrap-prime replied:
No, sorry. I can fix that.

bootstrap-prime commented:
Weird, ofborg is hanging but the logs look fine?

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/1008

SuperSandro2000 commented:
Please drop all the merge commits from this PR and rebase instead.

SuperSandro2000 commented on pkgs/development/python-modules/python-dbusmock/default.nix:
```diff
@@ -8,9 +8,15 @@
 }:
 
-buildPythonPackage rec {
+let
+  pbap-client = runCommand "pbap-client" { } ''
+    mkdir -p "$out/bin"
+    ln -s "${bluez.test}/test/pbap-client" "$out/bin/pbap-client"
+  '';
```
How about we just append ``${bluez.test}/test`` to PATH in preCheck? Thats faster and cleaner.

bootstrap-prime replied:
Sure, I can do that

bootstrap-prime replied:
Wait, now I remember why we can't do that. String substitution like that is constructed whether or not checkPhase is actually used. that unconditional inclusion introduces a dependency cycle when trying to add pcsclite as a dependency of libfido2.

jtojnar replied:
Forgot about that too, maybe we should add a comment about that so that we do not try it again in the future. Something like “Cannot just add it to path in `preCheck` since that attribute will be passed to `mkDerivation` even with `doCheck = false`, causing a dependency cycle.”

bootstrap-prime replied:
Good idea.

Mindavi commented:
3b9048a3cefb6e86af60348513624ea691cd7f41 broke cross-compilation because recursivePthLoader (which is only used for tests) doesn't cross-compile. Maybe that's the reason why it was conditional on doCheck? Now it's introduced in the closure for bluez and thus breaks a lot of downstream deps.

I can check if recursivePthLoader can be made to cross-compile properly, otherwise maybe we should revert that change that makes the test output unconditional?

cc @Artturin, maybe you have an idea?

```
patching sources
updateAutotoolsGnuConfigScriptsPhase
configuring
no configure script, doing nothing
building
/nix/store/4f9yfmxslh79x3v3zdnxy4hb2k1hw6zb-stdenv-linux/setup: line 1397: /nix/store/p3v2zvs7mgrf22c0wqv3vjjqvw1va9n6-python3-aarch64-unknown-linux-gnu-3.10.6/bin/python3.10: cannot execute binary file: Exec format error
```

Seems possibly an easy fix but out of time for today :)

Artturin commented:
> 3b9048a3cefb6e86af60348513624ea691cd7f41 broke cross-compilation because recursivePthLoader (which is only used for tests) doesn't cross-compile. Maybe that's the reason why it was conditional on doCheck? Now it's introduced in the closure for bluez and thus breaks a lot of downstream deps.
> 
> I can check if recursivePthLoader can be made to cross-compile properly, otherwise maybe we should revert that change that makes the test output unconditional?
> 
> cc @Artturin, maybe you have an idea?
> 
> ```
> patching sources
> updateAutotoolsGnuConfigScriptsPhase
> configuring
> no configure script, doing nothing
> building
> /nix/store/4f9yfmxslh79x3v3zdnxy4hb2k1hw6zb-stdenv-linux/setup: line 1397: /nix/store/p3v2zvs7mgrf22c0wqv3vjjqvw1va9n6-python3-aarch64-unknown-linux-gnu-3.10.6/bin/python3.10: cannot execute binary file: Exec format error
> ```
> 
> Seems possibly an easy fix but out of time for today :)



https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/python-modules/recursive-pth-loader/default.nix

It's using ${python} which is for host
It should be using ${python.pythonForBuild}

Mindavi commented:
See https://github.com/NixOS/nixpkgs/pull/185635 for someone working on a solution
